ring-io/src/ring.rs

```diff
@@ -0,0 +1,133 @@
+use crate::cq::CompletionQueue;
+use crate::register::Registrar;
+use crate::sq::SubmissionQueue;
+use crate::{sys, utils};
+
+use std::cell::UnsafeCell;
+use std::marker::PhantomData;
+use std::mem::MaybeUninit;
+use std::os::unix::io::RawFd;
+use std::ptr::NonNull;
+use std::{fmt, io, mem};
+
+#[repr(transparent)]
+pub(crate) struct RawRing(UnsafeCell<sys::io_uring>);
+
+impl RawRing {
+    pub unsafe fn new(entries: u32, params: &mut sys::io_uring_params) -> io::Result<Self> {
+        let mut ring: MaybeUninit<UnsafeCell<sys::io_uring>> = MaybeUninit::uninit();
+
+        let ring_ptr: *mut sys::io_uring = ring.as_mut_ptr().cast();
+
+        let ret = sys::io_uring_queue_init_params(entries, ring_ptr, params);
+        let _ = utils::resultify(ret)?;
+
+        Ok(Self(ring.assume_init()))
+    }
+
+    pub fn ring_fd(&self) -> RawFd {
+        unsafe { (*self.0.get()).ring_fd }
+    }
+
+    pub fn get_mut_ptr(&self) -> *mut sys::io_uring {
+        self.0.get()
+    }
+}
+
+pub(crate) struct RawRingPtr<'r>(NonNull<RawRing>, PhantomData<&'r mut RawRing>);
+
+impl RawRingPtr<'_> {
+    pub unsafe fn new_unchecked(ptr: *mut RawRing) -> Self {
+        Self(NonNull::new_unchecked(ptr), PhantomData)
+    }
+
+    pub fn get_ref(&self) -> &RawRing {
+        unsafe { self.0.as_ref() }
+    }
+
+    pub fn get_mut_ptr(&self) -> *mut sys::io_uring {
+        self.0.as_ptr().cast()
+    }
+}
+
+pub struct Ring {
+    ring: RawRing,
+}
+
+unsafe impl Send for Ring {}
+unsafe impl Sync for Ring {}
+
+pub struct RingBuilder {
+    entries: u32,
+    params: sys::io_uring_params,
+}
+
+unsafe impl Send for RingBuilder {}
+unsafe impl Sync for RingBuilder {}
+
+impl RingBuilder {
+    pub fn new(entries: u32) -> Self {
+        Self {
+            entries,
+            params: unsafe { mem::zeroed() },
+        }
+    }
+
+    pub fn build(mut self) -> io::Result<Ring> {
+        unsafe {
+            let ring = RawRing::new(self.entries, &mut self.params)?;
+            Ok(Ring { ring })
+        }
+    }
+}
```

Comment:
1. Why do we need a builder to get a ring?
2. We need to provide method to set ring setup params.

Reply:
Here is an example.

<https://github.com/tokio-rs/io-uring/blob/65daec6a7be843870166cbf549bc2f8a16198cba/src/lib.rs#L211-L304>

Reply:
So a user friend implementation for the 2nd question is good answer to the first one.